client/components/LinkPreview.vue

```diff
@@ -129,137 +129,203 @@
 	</div>
 </template>
 
-<script>
+<script lang="ts">
+import {
+	computed,
+	defineComponent,
+	inject,
+	nextTick,
+	onBeforeUnmount,
+	onMounted,
+	onUnmounted,
+	PropType,
+	ref,
+	watch,
+} from "vue";
+import {onBeforeRouteUpdate} from "vue-router";
 import eventbus from "../js/eventbus";
 import friendlysize from "../js/helpers/friendlysize";
+import {useStore} from "../js/store";
+import type {ClientChan, ClientLinkPreview} from "../js/types";
+import {imageViewerKey} from "./App.vue";
 
-export default {
+export default defineComponent({
 	name: "LinkPreview",
 	props: {
-		link: Object,
-		keepScrollPosition: Function,
-		channel: Object,
-	},
-	data() {
-		return {
-			showMoreButton: false,
-			isContentShown: false,
-		};
-	},
-	computed: {
-		moreButtonLabel() {
-			return this.isContentShown ? "Less" : "More";
+		link: {
+			type: Object as PropType<ClientLinkPreview>,
+			required: true,
 		},
-		imageMaxSize() {
-			if (!this.link.maxSize) {
+		keepScrollPosition: {
+			type: Function as PropType<() => void>,
+			required: true,
+		},
+		channel: {type: Object as PropType<ClientChan>, required: true},
+	},
+	setup(props) {
+		const store = useStore();
+
+		const showMoreButton = ref(false);
+		const isContentShown = ref(false);
+		const imageViewer = inject(imageViewerKey);
+
+		onBeforeRouteUpdate((to, from, next) => {
+			// cancel the navigation if the user is trying to close the image viewer
```

Comment:
opening does not push to history, and this correctly works on my device. 

This can't be done in closeViewer as onBeforeRouteUpdated needs to be done in a child of a `<router-view>` and the `<ImageViewer />` exists in `<App />` (as a sibling to the router-view)